File: logtui/src/editor.rs
use anyhow::Context;
use anyhow::Result;
use chrono::Local;
use std::env;
use std::fs;
use std::io::Write;
use std::process::Command;
use tempfile::NamedTempFile;

/// Get the editor command from environment or default to nvim
pub fn get_editor() -> String {
    let editor = env::var("EDITOR").unwrap_or_else(|_| "nvim".to_string());

    // Handle common variations
    match editor.as_str() {
        "neovim" => "nvim".to_string(),
        "vim" => "vim".to_string(),
        "vi" => "vi".to_string(),
        _ => editor,
    }
}

// Helper to YAML-quote a value when necessary. Placed at module scope so both
// edit_new_entry and edit_existing_entry can reuse it.
fn yaml_quote(s: &str) -> String {
    if s.is_empty() {
        return String::new();
    }
    // If value contains characters that require quoting, wrap in double quotes
    if s.starts_with(' ') || s.ends_with(' ') || s.contains(':') || s.contains('"') {
        let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
        return format!("\"{}\"", escaped);
    }
    s.to_string()
}

/// Open external editor for creating a new entry
/// Returns (location, content, tag, title) written by the user, or None if cancelled
pub fn edit_new_entry(
    location: Option<String>,
    default_tag: Option<String>,
    date: chrono::NaiveDate,
    _is_today: bool,
) -> Result<
    Option<(
        String,
        String,
        String,
        String,
        Option<String>,
        Option<String>,
    )>,
> {
    // Create temp file with .md extension
    let mut temp_file = NamedTempFile::with_suffix(".md").context("Failed to create temp file")?;

    let now = Local::now();

    // Helper to YAML-quote a value when necessary
    fn yaml_quote(s: &str) -> String {
        if s.is_empty() {
            return String::new();
        }
        // If value contains characters that require quoting, wrap in double quotes
        if s.starts_with(' ') || s.ends_with(' ') || s.contains(':') || s.contains('"') {
            let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
            return format!("\"{}\"", escaped);
        }
        s.to_string()
    }

    // Write YAML frontmatter
    writeln!(temp_file, "---")?;
    writeln!(temp_file, "date: {}", date.format("%Y-%m-%d"))?;
    writeln!(temp_file, "time: {}", now.time().format("%H:%M:%S"))?;
    // Title field for explicit title in frontmatter (empty by default)
    writeln!(temp_file, "title: {}", yaml_quote(""))?;
    // Write location even if empty; parser will accept empty location
    writeln!(
        temp_file,
        "location: {}",
        yaml_quote(&location.unwrap_or_default())
    )?;
    // Always include a tag field so user can edit it; default to 'log'
    let tag_default = default_tag.as_deref().unwrap_or("log");
    writeln!(temp_file, "tag: {}", yaml_quote(tag_default))?;
    writeln!(temp_file, "---")?;
    writeln!(temp_file)?;

    temp_file.flush()?;
    let temp_path = temp_file.path().to_path_buf();

    // Launch editor
    let editor = get_editor();
    let status = Command::new(&editor)
        .arg(&temp_path)
        .status()
        .with_context(|| format!("Failed to launch editor: {}", editor))?;

    if !status.success() {
        return Ok(None);
    }

    // Read back the content
    let content = fs::read_to_string(&temp_path).context("Failed to read temp file")?;

    // Parse YAML frontmatter and also return date/time from frontmatter
    if let Some((loc, body, tag, title)) = parse_yaml_frontmatter(&content)? {
        // Extract frontmatter date/time if present (fallbacks to provided values)
        // For now we return the date/time strings as written in frontmatter by
        // parsing the raw content: search for `date:` and `time:` lines.
        let mut fm_date = date.to_string();
        let mut fm_time = now.time().format("%H:%M:%S").to_string();
        for line in content.lines() {
            let t = line.trim();
            if t.starts_with("date:") {
                if let Some(v) = t.splitn(2, ':').nth(1) {
                    fm_date = v.trim().trim_matches('\"').to_string();
                }
            } else if t.starts_with("time:") {
                if let Some(v) = t.splitn(2, ':').nth(1) {
                    fm_time = v.trim().trim_matches('\"').to_string();
                }
            }
        }

        return Ok(Some((fm_date, fm_time, loc, body, tag, title)));
    }
    Ok(None)
}

/// Open external editor for editing an existing entry
/// Returns (date, time, day_of_week, location, content, tag, title) after editing, or None if cancelled
pub fn edit_existing_entry(
    date_str: &str,
    time_str: &str,
    _day_of_week: &str,
    location: &str,
    content: &str,
    tag: Option<&str>,
    title: Option<&str>,
) -> Result<
    Option<(
        String,
        String,
        String,
        String,
        Option<String>,
        Option<String>,
    )>,
> {
    // Create temp file with .md extension
    let mut temp_file = NamedTempFile::with_suffix(".md").context("Failed to create temp file")?;

    // reuse same yaml_quote helper

    // Write YAML frontmatter (include existing title line when editing)
    writeln!(temp_file, "---")?;
    writeln!(temp_file, "date: {}", date_str)?;
    writeln!(temp_file, "time: {}", time_str)?;
    // Prefill the title with the provided existing title (if any)
    writeln!(temp_file, "title: {}", yaml_quote(title.unwrap_or("")))?;
    writeln!(temp_file, "location: {}", location)?;
    if let Some(t) = tag {
        writeln!(temp_file, "tag: {}", t)?;
    }
    writeln!(temp_file, "---")?;
    writeln!(temp_file)?;
    writeln!(temp_file, "{}", content)?;

    temp_file.flush()?;
    let temp_path = temp_file.path().to_path_buf();

    // Launch editor
    let editor = get_editor();
    let status = Command::new(&editor)
        .arg(&temp_path)
        .status()
        .with_context(|| format!("Failed to launch editor: {}", editor))?;

    if !status.success() {
        return Ok(None);
    }

    // Read back the content
    let edited_content = fs::read_to_string(&temp_path).context("Failed to read temp file")?;

    // Parse YAML and return date/time extracted from frontmatter if present
    if let Some((loc, body, tag_res, title_res)) = parse_yaml_frontmatter(&edited_content)? {
        // Extract frontmatter date/time if present
        let mut fm_date = date_str.to_string();
        let mut fm_time = time_str.to_string();
        for line in edited_content.lines() {
            let t = line.trim();
            if t.starts_with("date:") {
                if let Some(v) = t.splitn(2, ':').nth(1) {
                    fm_date = v.trim().trim_matches('\"').to_string();
                }
            } else if t.starts_with("time:") {
                if let Some(v) = t.splitn(2, ':').nth(1) {
                    fm_time = v.trim().trim_matches('\"').to_string();
                }
            }
        }

        Ok(Some((fm_date, fm_time, loc, body, tag_res, title_res)))
    } else {
        Ok(None)
    }
}

/// Open external editor for editing a day's summary
/// Returns Some(text) if saved, None if cancelled
pub fn edit_summary(date: chrono::NaiveDate, current: &str) -> Result<Option<String>> {
    let mut temp_file = NamedTempFile::with_suffix(".md").context("Failed to create temp file")?;

    writeln!(temp_file, "# Summary for {}", date.format("%Y-%m-%d"))?;
    writeln!(temp_file)?;
    writeln!(temp_file, "{}", current)?;
    temp_file.flush()?;

    let temp_path = temp_file.path().to_path_buf();
    let editor = get_editor();
    let status = Command::new(&editor)
        .arg(&temp_path)
        .status()
        .with_context(|| format!("Failed to launch editor: {}", editor))?;

    if !status.success() {
        return Ok(None);
    }

    let content = std::fs::read_to_string(&temp_path).context("Failed to read temp file")?;
    // Return the whole file as the summary body (strip leading header line if present)
    let lines: Vec<&str> = content.lines().collect();
    let body = if !lines.is_empty() && lines[0].trim_start().starts_with('#') {
        lines
            .iter()
            .skip(1)
            .copied()
            .collect::<Vec<&str>>()
            .join("\n")
            .trim()
            .to_string()
    } else {
        content.trim().to_string()
    };

    // Return the body even if empty — caller decides whether empty means delete
    Ok(Some(body))
}

/// Parse the content from the editor, extracting tag from header
/// Returns (location, content, tag, title) if valid, None if cancelled
pub fn parse_yaml_frontmatter(
    content: &str,
) -> Result<Option<(String, String, Option<String>, Option<String>)>> {
    let lines: Vec<&str> = content.lines().collect();

    // Find YAML frontmatter delimiters more robustly (trimmed)
    if lines.is_empty() {
        return Ok(None);
    }

    let start_idx_opt = lines.iter().position(|l| l.trim() == "---");
    // If no frontmatter start marker, treat whole file as content (no metadata)
    if start_idx_opt.is_none() {
        let content_str = content.trim().to_string();
        if content_str.is_empty() {
            return Ok(None);
        }
        // No frontmatter: return location empty, content, no tag, no title
        return Ok(Some((String::new(), content_str, None, None)));
    }

    let start_idx = start_idx_opt.unwrap();
    // Find end marker after start
    let end_rel = lines
        .iter()
        .skip(start_idx + 1)
        .position(|l| l.trim() == "---");
    if end_rel.is_none() {
        return Ok(None);
    }
    let end_idx = start_idx + 1 + end_rel.unwrap();

    // Parse frontmatter
    let mut location: Option<String> = None;
    let mut tag: Option<String> = None;
    let mut title: Option<String> = None;

    // Helper to remove surrounding quotes and unescape common sequences
    fn unquote_value(v: &str) -> String {
        let s = v.trim();
        if s.len() >= 2 && s.starts_with('"') && s.ends_with('"') {
            // Remove surrounding double quotes and unescape \ and \"
            let inner = &s[1..s.len() - 1];
            let unescaped = inner.replace("\\\\", "\\").replace("\\\"", "\"");
            return unescaped;
        }
        if s.len() >= 2 && s.starts_with('\'') && s.ends_with('\'') {
            // YAML single-quote style: two single-quotes represent one
            let inner = &s[1..s.len() - 1];
            let unescaped = inner.replace("''", "'");
            return unescaped;
        }
        s.to_string()
    }

    for line in &lines[start_idx + 1..end_idx] {
        if let Some(colon_pos) = line.find(':') {
            let key = line[..colon_pos].trim();
            let value = line[colon_pos + 1..].trim();
            let v = unquote_value(value);

            match key {
                // Accept location even if empty string
                "location" => {
                    location = Some(v.clone());
                }
                "tag" => {
                    if !v.is_empty() {
                        tag = Some(v.clone());
                    }
                }
                "title" => {
                    // Allow empty title -> treat as None
                    if !v.is_empty() {
                        title = Some(v.clone());
                    }
                }
                "date" | "time" => {
                    // We read these but don't use them for new entries
                    // (we use current date/time instead)
                }
                _ => {}
            }
        }
    }

    // Get content after frontmatter
    let content_lines: Vec<&str> = lines.iter().skip(end_idx + 1).copied().collect();
    let content_str = content_lines.join("\n").trim().to_string();

    // Content may be empty when YAML frontmatter is present; accept empty content

    // Use empty string for missing location
    let final_location = location.unwrap_or_default();

    // DB only stores `tag` and now `title`
    let final_tag = tag;
    let final_title = title;

    Ok(Some((final_location, content_str, final_tag, final_title)))
}

File: tlog/editor.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

func getEditor() string {
	editor := os.Getenv("EDITOR")
	if editor == "" {
		editor = "nvim"
	}
	switch editor {
	case "neovim":
		return "nvim"
	default:
		return editor
	}
}

func yamlQuote(s string) string {
	if s == "" {
		return ""
	}
	if strings.HasPrefix(s, " ") || strings.HasSuffix(s, " ") || strings.Contains(s, ":") || strings.Contains(s, "\"") {
		escaped := strings.ReplaceAll(s, "\\", "\\\\")
		escaped = strings.ReplaceAll(escaped, "\"", "\\\"")
		return fmt.Sprintf("\"%s\"", escaped)
	}
	return s
}

// CreateNewEntryTempFile creates a temp file for a new entry and returns its path
func CreateNewEntryTempFile(location *string, defaultTag *string, date time.Time) (string, error) {
	tmpFile, err := os.CreateTemp("", "tlog-*.md")
	if err != nil {
		return "", fmt.Errorf("failed to create temp file: %w", err)
	}

	now := time.Now()
	loc := ""
	if location != nil {
		loc = *location
	}
	tag := "log"
	if defaultTag != nil {
		tag = *defaultTag
	}

	// Write YAML frontmatter
	fmt.Fprintln(tmpFile, "---")
	fmt.Fprintf(tmpFile, "date: %s\n", date.Format("2006-01-02"))
	fmt.Fprintf(tmpFile, "time: %s\n", now.Format("15:04:05"))
	fmt.Fprintf(tmpFile, "title: %s\n", yamlQuote(""))
	fmt.Fprintf(tmpFile, "location: %s\n", yamlQuote(loc))
	fmt.Fprintf(tmpFile, "tag: %s\n", yamlQuote(tag))
	fmt.Fprintln(tmpFile, "---")
	fmt.Fprintln(tmpFile)
	tmpFile.Close()

	return tmpFile.Name(), nil
}

// CreateEditEntryTempFile creates a temp file for editing an existing entry
func CreateEditEntryTempFile(dateStr, timeStr, location, content string, tag, title *string) (string, error) {
	tmpFile, err := os.CreateTemp("", "tlog-*.md")
	if err != nil {
		return "", fmt.Errorf("failed to create temp file: %w", err)
	}

	titleStr := ""
	if title != nil {
		titleStr = *title
	}

	// Write YAML frontmatter
	fmt.Fprintln(tmpFile, "---")
	fmt.Fprintf(tmpFile, "date: %s\n", dateStr)
	fmt.Fprintf(tmpFile, "time: %s\n", timeStr)
	fmt.Fprintf(tmpFile, "title: %s\n", yamlQuote(titleStr))
	fmt.Fprintf(tmpFile, "location: %s\n", location)
	if tag != nil {
		fmt.Fprintf(tmpFile, "tag: %s\n", *tag)
	}
	fmt.Fprintln(tmpFile, "---")
	fmt.Fprintln(tmpFile)
	fmt.Fprintln(tmpFile, content)
	tmpFile.Close()

	return tmpFile.Name(), nil
}

// CreateSummaryTempFile creates a temp file for editing a day's summary
func CreateSummaryTempFile(date time.Time, current string) (string, error) {
	tmpFile, err := os.CreateTemp("", "tlog-summary-*.md")
	if err != nil {
		return "", fmt.Errorf("failed to create temp file: %w", err)
	}

	fmt.Fprintf(tmpFile, "# Summary for %s\n\n", date.Format("2006-01-02"))
	fmt.Fprintln(tmpFile, current)
	tmpFile.Close()

	return tmpFile.Name(), nil
}

// EntryResult holds the parsed result from the editor
type EntryResult struct {
	Date     string
	Time     string
	Location string
	Content  string
	Tag      *string
	Title    *string
}

func parseYAMLFrontmatter(content, defaultDate, defaultTime string) (*EntryResult, error) {
	lines := strings.Split(content, "\n")
	if len(lines) == 0 {
		return nil, nil
	}

	// Find frontmatter delimiters
	startIdx := -1
	endIdx := -1
	for i, line := range lines {
		if strings.TrimSpace(line) == "---" {
			if startIdx == -1 {
				startIdx = i
			} else {
				endIdx = i
				break
			}
		}
	}

	if startIdx == -1 {
		// No frontmatter
		contentStr := strings.TrimSpace(content)
		if contentStr == "" {
			return nil, nil
		}
		return &EntryResult{
			Date:    defaultDate,
			Time:    defaultTime,
			Content: contentStr,
		}, nil
	}

	if endIdx == -1 {
		return nil, nil
	}

	result := &EntryResult{
		Date: defaultDate,
		Time: defaultTime,
	}

	// Parse frontmatter
	for _, line := range lines[startIdx+1 : endIdx] {
		colonIdx := strings.Index(line, ":")
		if colonIdx == -1 {
			continue
		}
		key := strings.TrimSpace(line[:colonIdx])
		value := unquoteValue(strings.TrimSpace(line[colonIdx+1:]))

		switch key {
		case "date":
			result.Date = value
		case "time":
			result.Time = value
		case "location":
			result.Location = value
		case "tag":
			if value != "" {
				result.Tag = &value
			}
		case "title":
			if value != "" {
				result.Title = &value
			}
		}
	}

	// Get content after frontmatter
	if endIdx+1 < len(lines) {
		result.Content = strings.TrimSpace(strings.Join(lines[endIdx+1:], "\n"))
	}

	return result, nil
}

func unquoteValue(v string) string {
	s := strings.TrimSpace(v)
	if len(s) >= 2 && strings.HasPrefix(s, "\"") && strings.HasSuffix(s, "\"") {
		inner := s[1 : len(s)-1]
		inner = strings.ReplaceAll(inner, "\\\\", "\\")
		inner = strings.ReplaceAll(inner, "\\\"", "\"")
		return inner
	}
	if len(s) >= 2 && strings.HasPrefix(s, "'") && strings.HasSuffix(s, "'") {
		inner := s[1 : len(s)-1]
		inner = strings.ReplaceAll(inner, "''", "'")
		return inner
	}
	return s
}

// Helper for reading user input
func promptChoice() (rune, error) {
	reader := bufio.NewReader(os.Stdin)
	line, err := reader.ReadString('\n')
	if err != nil {
		return 0, err
	}
	line = strings.TrimSpace(line)
	if len(line) == 0 {
		return '\n', nil
	}
	return rune(line[0]), nil
}
